include/zephyr/nvmem.h

```diff
@@ -0,0 +1,354 @@
+ *
+ * @see NVMEM_CELL_INST_GET_BY_IDX
+ */
+#define NVMEM_CELL_GET_BY_IDX(node_id, idx) NVMEM_CELL_INIT(DT_NVMEM_CELL_BY_IDX(node_id, idx))
+
+/**
+ * @brief Static initializer for a struct nvmem_cell from a DT_DRV_COMPAT
+ *        instance.
+ *
+ * @param inst DT_DRV_COMPAT instance number
+ * @param idx Logical index into 'nvmem-cells' property.
+ *
+ * @return Static initializer for a struct nvmem_cell for the property.
+ *
+ * @see NVMEM_CELL_GET_BY_IDX
+ */
+#define NVMEM_CELL_INST_GET_BY_IDX(inst, idx) NVMEM_CELL_GET_BY_IDX(DT_DRV_INST(inst), idx)
+
+/**
+ * @brief Like NVMEM_CELL_GET_BY_IDX(), with a fallback to a default value.
+ *
+ * If the devicetree node identifier 'node_id' refers to a node with a property
+ * 'nvmem-cells', this expands to <tt>NVMEM_CELL_GET_BY_IDX(node_id, idx)</tt>. The
+ * @p default_value parameter is not expanded in this case. Otherwise, this
+ * expands to @p default_value.
+ *
+ * @param node_id Devicetree node identifier.
+ * @param idx Logical index into 'nvmem-cells' property.
+ * @param default_value Fallback value to expand to.
+ *
+ * @return Static initializer for a struct nvmem_cell for the property,
+ *         or @p default_value if the node or property do not exist.
+ *
+ * @see NVMEM_CELL_INST_GET_BY_IDX_OR
+ */
+#define NVMEM_CELL_GET_BY_IDX_OR(node_id, idx, default_value)                                      \
+	COND_CODE_1(DT_NODE_HAS_PROP(node_id, nvmem_cells),                                        \
+		    (NVMEM_CELL_GET_BY_IDX(node_id, idx)),                                         \
+		    (default_value))
+
+/**
+ * @brief Like NVMEM_CELL_INST_GET_BY_IDX(), with a fallback to a default
+ *        value.
+ *
+ * @param inst DT_DRV_COMPAT instance number
+ * @param idx Logical index into 'nvmem-cells' property.
+ * @param default_value Fallback value to expand to.
+ *
+ * @return Static initializer for a struct nvmem_cell for the property,
+ *         or @p default_value if the node or property do not exist.
+ *
+ * @see NVMEM_CELL_GET_BY_IDX_OR
+ */
+#define NVMEM_CELL_INST_GET_BY_IDX_OR(inst, idx, default_value)                                    \
+	NVMEM_CELL_GET_BY_IDX_OR(DT_DRV_INST(inst), idx, default_value)
+
+/**
+ * @brief Read data from an NVMEM cell.
+ *
+ * @param cell The NVMEM cell.
+ * @param data Buffer to store read data.
+ * @param off The offset to start reading from.
+ * @param len Number of bytes to read.
+ *
+ * @kconfig_dep{CONFIG_NVMEM}
+ *
+ * @return 0 on success, negative errno code on failure.
+ */
+int nvmem_cell_read(const struct nvmem_cell *cell, void *data, off_t off, size_t len);
+
+/**
+ * @brief Write data to an NVMEM cell.
+ *
+ * @param cell The NVMEM cell.
+ * @param data Buffer with data to write.
+ * @param off The offset to start writing to.
+ * @param len Number of bytes to write.
+ *
+ * @kconfig_dep{CONFIG_NVMEM}
+ *
+ * @return 0 on success, negative errno code on failure.
```

Comment:
some of these use `@return` some use `@retval`, I vaugley remember that `@retval` needs multiple lines one per value, unsure on `@return` or if one should be used over the other